### PR TITLE
Feature 124 support weka

### LINKED ISSUE #124: Support for Weka
As a developer I would like to add support for Weka, because it contains an algorithm needed for recommendations.

### PR DESCRIPTION
**Pull request by platonovsimeon**
resolves #124

### CODE REVIEW

File: src/main/java/fi/metatavu/metamind/bot/script/BotWekaServiceScriptBinding.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	  if(!attributes.hasArrayElements()){
+	    return null;
+	  }
+	  int size = (int) attributes.getArraySize();
+	  String[] attrs = new String[size];
+	  for(int i = 0; i < size; i++) {
+	    attrs[i] = attributes.getArrayElement(i).asString();
+	  }
+	  Value ratingValue = item.getMember("rating");
+	  Double rating = ratingValue.fitsInDouble()?ratingValue.asDouble():null;
```
Comment (hkurhinen) [inline src/main/java/fi/metatavu/metamind/bot/script/BotWekaServiceScriptBinding.java]: Unnenessary check if rating fits in double, shoud be isNumber() also rating value should be checked for null value

File: src/main/java/fi/metatavu/metamind/weka/RecommendationData.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import weka.core.Instance;
+import weka.core.Instances;
+
+/**
+ * Helper class for converting data
+ * 
+ * @author Simeon Platonov
+ */
+public class RecommendationData {
+	ArrayList<WekaRecommendationItem> trainingItems;
```
Comment (hkurhinen) [inline src/main/java/fi/metatavu/metamind/weka/RecommendationData.java]: Indent is 4 and should be 2

File: src/main/java/fi/metatavu/metamind/weka/RecommendationData.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	 * @param items - All rated and unrated items
+	 */
+	public RecommendationData(WekaRecommendationItem[] items) {
+		this.items = items;
+	}
+	
+	/**
+	 * Converts text attributes to an integer array of 0s and 1s
+	 */
+	void convertAttributes() {
```
Comment (hkurhinen) [inline src/main/java/fi/metatavu/metamind/weka/RecommendationData.java]: Missing method access modifiers (in all methods of this class)

File: src/main/java/fi/metatavu/metamind/weka/RecommendationData.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	 */
+	void createDatasets() {
+		trainingSet = new Instances("trainingSet",attributeInfo,0);
+		recommendationSet = new Instances("recommendationSet",attributeInfo,0);
+		trainingSet.setClassIndex(attributeInfo.size()-1);
+		recommendationSet.setClassIndex(attributeInfo.size()-1);
+		
+		for(WekaRecommendationItem item:trainingItems) {
+			Instance instance = new DenseInstance(item.getConvertedAttributes().length+1);
+			for(int i=0;i<item.getConvertedAttributes().length;i++) {
```
Comment (hkurhinen) [inline src/main/java/fi/metatavu/metamind/weka/RecommendationData.java]: add spaces to the for loop conditions eq. int i=0; i<item.getConvertedAttributes().length; i++

(In other for loops also)

File: src/main/java/fi/metatavu/metamind/weka/RecommendationData.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+import weka.core.DenseInstance;
+import weka.core.Instance;
+import weka.core.Instances;
+
+/**
+ * Helper class for converting data
+ * 
+ * @author Simeon Platonov
+ */
+public class RecommendationData {
```
Comment (hkurhinen) [inline src/main/java/fi/metatavu/metamind/weka/RecommendationData.java]: Class naming is bit confusing, better would be e.q. RecommendationDataUtils

File: src/main/java/fi/metatavu/metamind/weka/WekaController.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import weka.core.Instance;
+
+
+/**
+ * Functionality from Weka framework
+ * 
+ * @author Simeon Platonov
+ */
+public class WekaController {
+	public WekaController() {
```
Comment (hkurhinen) [inline src/main/java/fi/metatavu/metamind/weka/WekaController.java]: indent is 4 and should be 2

File: src/main/java/fi/metatavu/metamind/weka/WekaController.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+		data.convertAttributes();
+		data.splitData();
+		data.createAttributeInfo();
+		data.createDatasets();
+		LinearRegression model = new LinearRegression();
+		try {
+			model.buildClassifier(data.trainingSet);
+		} catch (Exception e1) {
+			// TODO Auto-generated catch block
+			e1.printStackTrace();
```
Comment (hkurhinen) [inline src/main/java/fi/metatavu/metamind/weka/WekaController.java]: Use logger and log error properly

File: src/main/java/fi/metatavu/metamind/weka/WekaController.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+		}
+		ArrayList<WekaRecommendationItem> recommendedItems = new ArrayList<WekaRecommendationItem>();
+		for(int i = 0;i<data.recommendationSet.size();i++) {
```
Comment (hkurhinen) [inline src/main/java/fi/metatavu/metamind/weka/WekaController.java]: fix formatting in for loop

File: src/main/java/fi/metatavu/metamind/weka/WekaController.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+				
+					try {
+						double estimatedRating = recommend(data.recommendationSet.get(i),model);
+						WekaRecommendationItem item = data.itemsToRecommend.get(i);
+						item.setRating(estimatedRating);
+						recommendedItems.add(item);
+					} catch (Exception e) {
+						// TODO Auto-generated catch block
+						e.printStackTrace();
```
Comment (hkurhinen) [inline src/main/java/fi/metatavu/metamind/weka/WekaController.java]: Use logger and log error properly

File: src/main/java/fi/metatavu/metamind/weka/WekaRecommendationItem.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * Class for recommendation items
+ * 
+ * @author Simeon Platonov
+ */
+public class WekaRecommendationItem implements Comparable<WekaRecommendationItem>{
+  private Double rating = null;
+  private String id;
+	private String[] attributes;
```
Comment (hkurhinen) [inline src/main/java/fi/metatavu/metamind/weka/WekaRecommendationItem.java]: fix formatting

File: src/main/java/fi/metatavu/metamind/weka/WekaRecommendationItem.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+	}
+	public void setId(String id) {
+    this.id=id;
+  }
+	public Double getRating() {
+    return this.rating;
+  }
+
+  public void setRating(Double rating) {
+
```
Comment (hkurhinen) [inline src/main/java/fi/metatavu/metamind/weka/WekaRecommendationItem.java]: Remove unnecessary whitespace

File: .metadata/.log [hunk 12]
```diff
@@ -0,0 +1,10 @@
+!SESSION 2019-06-11 17:52:58.743 -----------------------------------------------
```
[the remark this answers is not in the document]
Comment (platonovsimeon) [reply .metadata/.log]: Removed

File: pom.xml [hunk 13]
```diff
@@ -480,8 +480,16 @@
       <scope>test</scope>
     </dependency>
     
+    <!-- Weka -->
+     <dependency>
+       <groupId>nz.ac.waikato.cms.weka</groupId>
+       <artifactId>weka-stable</artifactId>
+       <version>3.8.3</version>
```
[the remark this answers is not in the document]
Comment (platonovsimeon) [reply pom.xml]: Fixed

File: pom.xml [hunk 14]
```diff
@@ -480,8 +480,16 @@
...
+	</dependency>
```
[the remark this answers is not in the document]
Comment (platonovsimeon) [reply pom.xml]: Fixed

File: src/main/java/fi/metatavu/metamind/bot/script/BotWekaServiceScriptBinding.java [hunk 15]
```diff
@@ -0,0 +1,83 @@
+package fi.metatavu.metamind.bot.script;
+
+
+
```
[the remark this answers is not in the document]
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/bot/script/BotWekaServiceScriptBinding.java]: Removed

File: src/main/java/fi/metatavu/metamind/weka/RecommendationData.java (hunk 5 above)
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/weka/RecommendationData.java]: Class naming changed

File: src/main/java/fi/metatavu/metamind/weka/RecommendationData.java (hunk 3 above)
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/weka/RecommendationData.java]: Access modifiers added

File: src/main/java/fi/metatavu/metamind/weka/WekaRecommendationItem.java (hunk 11 above)
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/weka/WekaRecommendationItem.java]: Removed

File: src/main/java/fi/metatavu/metamind/bot/script/BotWekaServiceScriptBinding.java [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+import fi.metatavu.metamind.weka.WekaRecommendationItem;
+
+/**
+ * Bot script binding for Weka service
+ * 
+ * @author Simeon Platonov
+ */
+@ApplicationScoped
+public class BotWekaServiceScriptBinding {
+
```
[the remark this answers is not in the document]
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/bot/script/BotWekaServiceScriptBinding.java]: Indents fixed

File: src/main/java/fi/metatavu/metamind/weka/WekaRecommendationItem.java (hunk 10 above)
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/weka/WekaRecommendationItem.java]: Fixed formatting

File: src/main/java/fi/metatavu/metamind/weka/WekaController.java (hunk 8 above)
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/weka/WekaController.java]: Fixed

File: src/main/java/fi/metatavu/metamind/weka/WekaController.java (hunk 6 above)
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/weka/WekaController.java]: Fixed

File: src/main/java/fi/metatavu/metamind/weka/RecommendationData.java (hunk 2 above)
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/weka/RecommendationData.java]: Fixed indents

File: src/main/java/fi/metatavu/metamind/weka/WekaController.java (hunk 7 above)
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/weka/WekaController.java]: Fixed

File: src/main/java/fi/metatavu/metamind/weka/WekaController.java (hunk 9 above)
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/weka/WekaController.java]: Fixed

File: src/main/java/fi/metatavu/metamind/weka/RecommendationData.java (hunk 4 above)
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/weka/RecommendationData.java]: Fixed for-loop formatting

File: src/main/java/fi/metatavu/metamind/bot/script/BotWekaServiceScriptBinding.java [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+	      .map(this::translateWekaRecommendationItem)
+	      .filter(Objects::nonNull)
+	      .toArray(WekaRecommendationItem[]::new);
+	  
+	  String[] recommendations = wekaController.getRecommendations(wekaItems);
+	  return recommendations;
+	}
+	
+	private WekaRecommendationItem translateWekaRecommendationItem(Value item) {
+	  String id = item.getMember("id").asString();
```
[the remark this answers is not in the document]
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/bot/script/BotWekaServiceScriptBinding.java]: Fixed

File: src/main/java/fi/metatavu/metamind/bot/script/BotWekaServiceScriptBinding.java [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+	  Value attributes = item.getMember("attributes");
+	  if(!attributes.hasArrayElements()){
```
[the remark this answers is not in the document]
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/bot/script/BotWekaServiceScriptBinding.java]: Fixed

File: src/main/java/fi/metatavu/metamind/bot/script/BotWekaServiceScriptBinding.java (hunk 1 above)
Comment (platonovsimeon) [reply src/main/java/fi/metatavu/metamind/bot/script/BotWekaServiceScriptBinding.java]: Check changed
